### ISSUE TITLE
Help with slow SceneCanvas use case

### ISSUE THREAD

**Issue by djhoese** [vispy/vispy#1826]
I just found the cause to my problem in this topic. I'm trying to create a graphical display of a mechanical system (suspension system). It basically contains cylinders and spheres, but it ran so slowly that I came to think it was a problem of too much data points (wasn't even 1M). Then I tried to use line plots (LinePlot3D tutorial) and the problem persists. Is there any way I can do it, in a way that every part works independently? A picture of what I have now follows:

![image](https://user-images.githubusercontent.com/59147674/75079379-46c95680-54e7-11ea-891c-72aa91d40405.png)

I fit into the 'scientists with no prior knowledge of OpenGL' group, btw.

_Originally posted by @ariel-avi in https://github.com/vispy/vispy/issues/1700#issuecomment-589877588_

**Comment by djhoese** [comment]
I'm moving this issue/comment from @ariel-avi here so we can have a more in depth conversation. @ariel-avi, it is difficult to say exactly what could be improved without seeing the code for your current solution. Could you provide a minimal example that shows your issue (something that I could run without having to download data files)?

In general you want to avoid making a separate Visual for each object in your visualization. If you can combine as many objects as possible in to one Visual then you'll get much better performance.

**Comment by ariel-avi** [comment]
Hello David, thank you for the help. You can run the code below. The idea is to have separate objects (in this case, wishbones, but it can be any other body) whose positions in space are to be individually manipulated (for multibody analysis) and overlaid to compare. Is there any way I can do this in a efficient way? How can I make one visual with multiple objects?

The plots must remain interactive, but I'd also like to add a little coordinate system on the bottom left of the window.

```
def draw_system():

    from PyQt5.QtWidgets import QWidget, QGridLayout, QMainWindow
    from PyQt5.Qt import QApplication
    from vispy import scene, visuals
    import numpy as np

    hard_points_front = {
        'lca_fore': np.array([167.31, -200., 165.87]),
        'lca_aft': np.array([-169.52, -200., 190.14]),
        'lca_outer': np.array([9., -565.82, 157.92]),
        'uca_fore': np.array([134.02, -232.51, 325.52]),
        'uca_aft': np.array([-131.26, -232.51, 325.52]),
        'uca_outer': np.array([-9., -533.29, 367.92]),
    }

    hard_points_rear = {
        'lca_fore': np.array([1426.5, -295.51, 210.66]),
        'lca_aft': np.array([1636.06, -298.62, 193.47]),
        'lca_outer': np.array([1601., -573.25, 177.994]),
        'uca_fore': np.array([1430.73, -277.68, 320.41]),
        'uca_aft': np.array([1644.2, -281.05, 305.93]),
        'uca_outer': np.array([1601., -543.29, 370.014])
    }

    class SystemsViewCanvas(scene.SceneCanvas):

        Draw3D = scene.visuals.create_visual_node(visuals.LinePlotVisual)

        def __init__(self, parent: QWidget = None, label: str = '', size: tuple = ('width', 'height')):
            if all(isinstance(i, str) for i in size):
                size = (800, 600)
            else:
                if parent is not None:
                    size = (parent.width(), parent.height())
            super(SystemsViewCanvas, self).__init__(title='', size=size)
            self.unfreeze()
            self.parent = parent
            self.bgcolor = [1, 1, 1]
            self._view = self.central_widget.add_view()
            self._view.camera = scene.TurntableCamera()
            # self.show()
            self.label = scene.visuals.Text(label,
                                            parent=self.scene,
                                            pos=(10, 10),
                                            font_size=8,
                                            face='Courier New',
                                            anchor_x='left',
                                            anchor_y='bottom',
                                            method='gpu')

        def draw_system(self):

            """

            Draws the system

            :return: None

            """

            _wishbone = dict()
            for axle in (hard_points_front, hard_points_rear):
                for side in {'left', 'right'}:
                    for prefix in {'uca', 'lca'}:
                        for suffix in {'_fore', '_aft'}:
                            # Draw wishbones
                            key = prefix + suffix
                            if side == 'left':
                                point = np.array([axle[key]*[1, -1, 1], axle[prefix + '_outer']*[1, -1, 1]])
                            else:
                                point = np.array([axle[key], axle[prefix + '_outer']])
                            _wishbone[prefix] = self.Draw3D(point,
                                                            width=2.0,
                                                            color='red',
                                                            edge_color='w',
                                                            symbol='o',
                                                            face_color=(0.2, 0.2, 1, 0.8),
                                                            parent=self._view.scene)

        @classmethod
        def isometric_view(cls, parent=None, size=('width', 'height')):
            canvas = cls(parent, 'Isometric view', size)
            canvas._view.camera = scene.TurntableCamera(30., azimuth=-45.)
            canvas._view.camera.set_range((-200, 1800), (-800, 800), (-100, 800), )
            return canvas

        @classmethod
        def front_view(cls, parent=None, size=('width', 'height')):
            canvas = cls(parent, 'Front View', size)
            canvas._view.camera = scene.PanZoomCamera(aspect=1., up='+x')
            canvas._view.camera.set_range(y=(-600, 600), z=(0, 500))
            canvas._view.camera.pan((0, 300))
            canvas._view.camera.zoom(.75)
            return canvas

        @classmethod
        def side_view(cls, parent=None, size=('width', 'height')):
            canvas = cls(parent, 'Side View', size)
            canvas._view.camera = scene.PanZoomCamera(aspect=1., up='+y')
            canvas._view.camera.set_range(x=(-200, 1800), z=(0, 800))
            canvas._view.camera.pan((0, 300))
            return canvas

        @classmethod
        def top_view(cls, parent=None, size=('width', 'height')):
            canvas = cls(parent, 'Top View', size)
            canvas._view.camera = scene.PanZoomCamera(aspect=1., up='+z')
            canvas._view.camera.set_range(x=(-200, 1800), y=(-800, 800))
            return canvas

    class GraphicsTab(QWidget):

        """
        Sets up the central widget
        """

        def __init__(self, parent=None):
            super(GraphicsTab, self).__init__(parent)
            self._current_project = None
            self.setLayout(QGridLayout())
            self.iso_view = SystemsViewCanvas.isometric_view(self)
            self.front_view = SystemsViewCanvas.front_view(self)
            self.side_view = SystemsViewCanvas.side_view(self)
            self.top_view = SystemsViewCanvas.top_view(self)
            self.views = (self.iso_view, self.front_view, self.side_view, self.top_view)
            i, j = 0, 2
            for view in self.views:
                self.layout().addWidget(view.native, i // j, i % j)
                i += 1

        def draw_system(self):
            # this code simulates 3 systems overlapped (for comparison reasons)
            for i in range(3):
                self.iso_view.draw_system()
                self.front_view.draw_system()
                self.side_view.draw_system()
                self.top_view.draw_system()

    app = QApplication([])
    window = QMainWindow()
    window.setCentralWidget(GraphicsTab())
    window.show()
    window.centralWidget().draw_system()
    app.exec()


if __name__ == '__main__':
    draw_system()

```

**Comment by djhoese** [comment]
A couple questions:

1. Is there a reason you declare your classes inside a function? It isn't wrong, but I've just never seen it done this way before.
2. Is there a reason you do `Draw3D = scene.visuals.create_visual_node(visuals.LinePlotVisual)` instead of using `vispy.scene.visuals.LinePlot`? Your code is the same as what's in VisPy so you probably don't need this extra code.

If my math is correct you are creating 16 LinePlot visuals, which themselves are CompoundVisual objects made up of a LineVisual and a MarkersVisual. Then you are creating 4 views. So thats 16 * 2 * 4 = 128 visuals. The issue here is that each of these visuals is a separate OpenGL program. Your GPU runs each of these program serially (one at a time). So that means it has to run 128 programs for each frame which will be slow.

The best approach in the long run is to combine as much of your logic in to a few GL programs as possible. For example, you have one Visual for each line segment. LineVisuals are able to draw millions of line segments efficiently if needed. Similarly Markers can show a lot of logically separate markers in one single Visual object. I would suggest taking a look at the docstrings for these Visuals and see if it gives you some ideas for how you could structure the data you are providing to VisPy and your GPU. You may need to create your own LinePlotVisual class to implement things the way you need since the builtin LinePlotVisual likely doesn't make all of these features easily accessible to the user (to keep interfaces simple).

Lastly, VisPy used to allow for multiple "views" of the same set of visuals. I'm 99% this was broken after a couple refactors back in VisPy 0.4, but it is something I want to add back in. However, it would require a lot of work. I just wanted to mention it.

Hopefully this clears a few things up. I'll try to explain as much as I can if you have more questions.

**Comment by ariel-avi** [comment]
The answer for your questions:

1. The reason to declare a class inside a function is that I use a single file to run code examples and it is easier to debug and keeps code organized.
2. I used the same code in vispy example. It is easier to read, also.

Am I able to move objects separately inherited in a compound visual? If it is, I may think how I'll overcome my problem.

**Comment by djhoese** [comment]
> The reason to declare a class inside a function is that I use a single file to run code examples and it is easier to debug and keeps code organized.

Interesting. I've never seen this pattern before. I guess I would expect one example per python module. There is no reason the classes can't be defined in the global/module namespace and a single function (ex. `main`) that creates the various objects and runs things. I would think this is better for debugging as the debugger has access to the class in the global namespace. Whatever works for you. I was just curious.

> I used the same code in vispy example. It is easier to read, also.

I meant you could replace:

```
Draw3D = scene.visuals.create_visual_node(visuals.LinePlotVisual)
```

with

```
Draw3D = scene.visuals.LinePlot
```

And then you could remove the `visuals` import you have now. Or you could do `from vispy.scene.visuals import LinePlot as Draw3D` if you really wanted to.

Anyway, as for your question:

> Am I able to move objects separately inherited in a compound visual?

I'm not sure I know what you mean. Handling click and drag functionality will be equally difficult (I think) for CompoundVisual versus separate visuals. You get a mouse event, you figure out which visual element was clicked, then do `set_data` on the line and the markers to update their positions.  Picking out one point/line from a single Visual can be difficult, but there should be examples for this. If not in the VisPy examples, it has been discussed in our issues before.

**Comment by ariel-avi** [comment]
Ah, now I get it foe the Draw3D, sorry I didn't understand quite well what you meant.

About the CompoundVisual, I came to think (probably equivocally) that the visuals that build a CompoundVisual are "frozen" together and couls only move all together. Please correct me if I'm wrong.

If I'm right about the functionality of CompoundVisual class, then it's a dead end for me to use Visuals on my application.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________

**Comment by ariel-avi** [comment]
Ah, now I get it for the Draw3D, sorry I didn't understand quite well what you meant.

About the CompoundVisual, I came to think (probably equivocally) that the visuals that build a CompoundVisual are "frozen" together and couls only move all together. Please correct me if I'm wrong.

If I'm right about the functionality of CompoundVisual class, then it's a dead end for me to use Visuals on my application.

**Comment by djhoese** [comment]
A `CompoundVisual` (or rather, a subclass of the `CompoundVisual` class) is a simple wrapper around two or more other Visual objects. You can view the source code for the LinePlotVisual in the vispy code and notice that it is rather simple by itself:

https://github.com/vispy/vispy/blob/master/vispy/visuals/line_plot.py

A lot of that code is just making sure that both visuals are updated at the same time. The `CompoundVisual` base class is a simple wrapper making sure that both visuals are redrawn when one of them is updated. There is no reason you can't make your own subclass of the CompoundVisual if you'd like and do whatever you need to inside it. Another option would be to create the individual LineVisual and MeshVisual in your Canvas and coordinate things yourself (the SceneCanvas does most of this automatically for you).

Bottom line, the CompoundVisual is a convenience but not something you need to use.

**Comment by ariel-avi** [comment]
Hm, I will need to review documentation to understand how it works before commenting anything again.

Thank you so much for helping me out so far.

**Comment by larsoner** [comment]
Closing as this is hopefully solved or covered by existing performance-issue-issues
